### PR TITLE
Added scan operation in dygo

### PR DESCRIPTION
**Pull request by Dinesh-Quollio**
Added scan operation in dygo.

[TICKET](https://linear.app/quollio/issue/SAAS-1971/add-scan-operation-in-dygo)

### CODE REVIEW

**Review by zephyrchien — APPROVED** [review]
LGTM!

File: scan.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	output, err := i.c.client.Scan(ctx, input)
+	if err != nil {
+		if err := getDynamoDBError(opScan, err); err != nil {
+			return nil, err
+		}
+		return nil, dynamoError().method(opScan).message(err.Error())
+	}
+
+	result.Results = output.Items
+	result.LastEvaluatedKey = output.LastEvaluatedKey
```
Comment (zephyrchien) [inline scan.go:104]: BTW could we also have this for `Query`

Comment (Dinesh-Quollio) [reply scan.go:104]: Sure, created [ticket](https://linear.app/quollio/issue/SAAS-1976/return-lastevaluatedkey-for-query-operation) for that.

File: scan.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+func (i *Item) scanWithLimit(ctx context.Context, input *dynamodb.ScanInput, result *output) (*output, error) {
+	output, err := i.c.client.Scan(ctx, input)
+	if err != nil {
+		if err := getDynamoDBError(opScan, err); err != nil {
+			return nil, err
+		}
+		return nil, dynamoError().method(opScan).message(err.Error())
+	}
+	// Handle pagination limits
+	if i.pagination.limit > 0 {
```
Comment (jdzq) [inline scan.go:83]: Hmm, but this works on data that was already fetched from DDB. 

We then fetch everything starting from LEK and throw away anything that's above the limit. On every page we would be fetching everything and not using it. 🙈

We can instead pass the `limit` to `scan_input`, no? ... and if we do it then the DDB will return LEK in resp.

We then potentially don't need `scan` and `scanWithLimit` - they could be the same.

Comment (Dinesh-Quollio) [reply scan.go]: @jdzq : Thank you for your comment.

The reason for not using the Limit parameter in the scan_input is due to how DynamoDB behaves when both a filter and limit are applied in a scan operation. DynamoDB fetches the specified Limit (e.g., 100 items) from the table first and then applies the filter on those items. This means that if many of the fetched items don’t match the filter criteria, we could end up with very few (or even no) results in a single scan operation.

By explicitly handling the limit ourselves, we can reduce the number of calls to DynamoDB, as we control the logic to aggregate results across pages. This allows us to fetch all relevant items efficiently, ensuring we only stop when we have the required number of filtered items.

**Here is an example (test-case)-**

**Example: Filtering with and without Limit in scan_input**
We have 5,150 items in our table, out of which only 150 items are of interest (i.e., they match the filter criteria). The limit is set to 250.

**Case 1: Using Limit in scan_input**
When the Limit parameter is applied in scan_input, DynamoDB retrieves 250 items at a time and then applies the filter. As a result, we may need to make 21 scan calls to fetch all 150 items. Here's a breakdown of the number of filtered items returned per page:

Items returned per page after applying the filter:

> Page 1: 3
> Page 2: 12
> Page 3: 8
> Page 4: 3
> Page 5: 10
> Page 6: 12
> Page 7: 7
> Page 8: 7
> Page 9: 8
> Page 10: 6
> Page 11: 7
> Page 12: 6
> Page 13: 4
> Page 14: 5
> Page 15: 17
> Page 16: 6
> Page 17: 6
> Page 18: 3
> Page 19: 6
> Page 20: 8
> Page 21: 6

```
Total items returned: 150
Number of scan calls required: 21
```

This approach fetches a large number of items (250 per page), but many are discarded after applying the filter, leading to inefficiency.


**Case 2: Explicitly Handling the Limit**
If we control the logic and explicitly set a custom limit, we can reduce the number of scan calls. For instance, in this example, we only need 2 scan calls to fetch all 150 items:

Items returned per page after applying the filter:

> Page 1: 139
> Page 2: 11

```
Total items returned: 150
Number of scan calls required: 2
```


[I have slightly modified scan logic to optimize it.]

Comment (jdzq) [reply scan.go]: Yeah, the scan limit has a very narrow use case where it's useful. Here example:
For instance (using data from your test case) if we wanted to fetch only 3 items we could set the limit to 250 and pay only for 250 scans to get the 3 items. ... but in the current implementation we don't pass the limit to DDB so we would pay for maximum number of scans per request and we would still return only the 3 items after truncating the response which would be **_at least_ 10 times more expensive** (as per your test case 2 requests scan 5k items so one of them scans at minimum 2.5k). However this is a rare use case. 

I understand that the introduced limit is supposed to control the page size which is different the DDB scan limit.

**Review by jdzq — APPROVED** [review]

**Comment by jdzq** [comment]
+1 for introducing the paginator so we don't need to handle that logic in `qdc-backends`.
